plugins/inputs/cloud_pubsub/pubsub.go

```diff
@@ -119,13 +127,46 @@ func (ps *PubSub) Stop() {
 	ps.wg.Wait()
 }
 
-func (ps *PubSub) subReceive(cctx context.Context) {
+// startReceiver is called within a goroutine and manages keeping a
+// subscription.Receive() up and running while the plugin has not been stopped.
+func (ps *PubSub) receiveWithRetry() {
+	err := ps.startReceiver()
+
+	for err != nil && ps.ctx.Err() == nil {
+		log.Printf("E! Receiver for subscription %s exited with error: %v", ps.sub.ID(), err)
+
+		delay := defaultRetryDelaySeconds
+		if ps.RetryReceiveDelaySeconds > 0 {
+			delay = ps.RetryReceiveDelaySeconds
+		}
+
+		log.Printf("I! Waiting %d seconds before attempting to restart receiver...", delay)
+		time.Sleep(time.Duration(delay) * time.Second)
+
+		err = ps.startReceiver()
+	}
+}
+
+func (ps *PubSub) startReceiver() error {
+	cctx, ccancel := context.WithCancel(ps.ctx)
```

Comment:
Move the creation of the cancel function into receiveOnSub, and just pass the context down.